packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/sign_out_state_machine.dart

```diff
@@ -0,0 +1,189 @@
+
+  /// The Cognito user pool configuration.
+  CognitoUserPoolConfig get _userPoolConfig => expect();
+
+  Future<void> _onInitiate(SignOutInitiate event) async {
+    final options = event.options;
+
+    // Try to retrieve tokens and return successfully if already logged out.
+    // Do not clear other storage items (e.g. AWS credentials) in this case,
+    // since an unauthenticated user may still be cached.
+    final CognitoUserPoolTokens tokens;
+    try {
+      tokens = await manager.getUserPoolTokens();
+    } on SignedOutException {
+      return emit(
+        const SignOutState.complete(CognitoSignOutResult.complete()),
+      );
+    }
+
+    // Capture results of individual steps to determine overall success.
+    HostedUiException? hostedUiException;
+    GlobalSignOutException? globalSignOutException;
+    RevokeTokenException? revokeTokenException;
+
+    // Sign out via Hosted UI, if configured.
+    Future<CognitoSignOutResult?> signOutHostedUi() async {
+      if (tokens.signInMethod == CognitoSignInMethod.hostedUi) {
+        final hostedUiResult = await manager.signOutHostedUI();
+        if (hostedUiResult is HostedUiFailure) {
+          final exception = hostedUiResult.exception;
+          if (exception is UserCancelledException) {
+            return CognitoSignOutResult.failed(exception);
+          }
+          hostedUiException = HostedUiException(
+            underlyingException: hostedUiResult.exception,
+          );
+        }
+      }
+      return null;
+    }
+
+    // On native platforms, Hosted UI should be logged out first. This gives
+    // users the opportunity to cancel the sign out flow if they wish before
+    // credentials are revoked and cleared.
+    //
+    // On Web, this should be the very last thing to happen. Since we redirect
+    // as a result of signing out Hosted UI, credentials should be revoked and
+    // cleared before this happens.
+    if (!zIsWeb) {
+      final hostedUiResult = await signOutHostedUi();
+      if (hostedUiResult != null) {
+        return emit(SignOutState.complete(hostedUiResult));
+      }
+    }
+
+    // Do not try to send Cognito requests for plugin configs without an
+    // Identity Pool, since the requests will fail.
+    if (_identityPoolConfig != null) {
+      if (options.globalSignOut) {
+        // Revokes the refresh token
+        try {
+          await _cognitoIdp
+              .globalSignOut(
+                GlobalSignOutRequest(
+                  accessToken: tokens.accessToken.raw,
+                ),
+              )
+              .result;
+        } on Exception catch (e) {
+          globalSignOutException = GlobalSignOutException(
+            accessToken: tokens.accessToken.raw,
+            underlyingException: e,
+          );
+        }
+      }
+      // Revokes the access token
+      if (globalSignOutException != null) {
+        revokeTokenException = RevokeTokenException(
+          refreshToken: tokens.refreshToken,
```

Comment:
what is the need to capture refreshToken in this exception?

Reply:
It gives users the opportunity to retry the revoke on their own. After this function completely, all the tokens are cleared locally.